spotfinder/spotfinder.cu

```diff
@@ -9,118 +9,278 @@
+     * from the background calculation in the second pass.
+    */
+    PitchedMalloc<uint8_t> d_erosion_mask(width, height);
+
+    {  // Scope for erosion pass launch parameters
+        // dim3 threads_per_erosion_block(32, 32);
+        // dim3 erosion_blocks(
+        //   (width + threads_per_erosion_block.x - 1) / threads_per_erosion_block.x,
+        //   (height + threads_per_erosion_block.y - 1) / threads_per_erosion_block.y);
+
+        // Calculate the shared memory size for the erosion kernel
+        // size_t erosion_shared_memory =
+        //   (threads_per_erosion_block.x + 2 * first_pass_kernel_radius)
+        //   * (threads_per_erosion_block.y + 2 * first_pass_kernel_radius)
+        //   * sizeof(uint8_t);
+
+        // Perform erosion
+        erosion_kernel<<<blocks, threads, shared_memory, stream>>>(
+          d_dispersion_mask.get(),
+          d_erosion_mask.get(),
+          mask.get(),
+          d_dispersion_mask.pitch,
+          d_erosion_mask.pitch,
+          mask.pitch,
+          width,
+          height,
+          first_pass_kernel_radius);
+        cudaStreamSynchronize(
+          stream);  // Synchronize the CUDA stream to ensure the erosion pass is complete
+
+        if (do_writeout) {
+            printf("Erosion pass complete\n");
+            debug_writeout(
+              d_erosion_mask.get(),
+              d_erosion_mask.pitch_bytes(),
+              width,
+              height,
+              stream,
+              "eroded_dispersion_result",
+              [](uint8_t pixel) { return pixel == MASKED_PIXEL ? 0 : 255; },
+              [](uint8_t pixel) { return pixel == MASKED_PIXEL; });
         }
     }
 
-    if (x < width && y < height) {
-        // result_sum[x + image_pitch * y] = sum;
-        // result_sumsq[x + image_pitch * y] = sumsq;
-        // result_n[x + mask_pitch * y] = n;
-
-        // Calculate the thresholding
-        if (px_is_valid) {
-            constexpr float n_sig_s = 3.0f;
-            constexpr float n_sig_b = 6.0f;
-
-            float sum_f = static_cast<float>(sum);
-            float sumsq_f = static_cast<float>(sumsq);
-
-            float mean = sum_f / n;
-            float variance = (n * sumsq_f - (sum_f * sum_f)) / (n * (n - 1));
-            float dispersion = variance / mean;
-            float background_threshold = 1 + n_sig_b * sqrt(2.0f / (n - 1));
-            bool not_background = dispersion > background_threshold;
-            float signal_threshold = mean + n_sig_s * sqrt(mean);
-            bool is_signal = this_pixel > signal_threshold;
-            bool is_strong_pixel = not_background && is_signal;
-            result_strong[x + mask_pitch * y] = is_strong_pixel;
-        } else {
-            result_strong[x + mask_pitch * y] = 0;
+    constexpr uint8_t second_pass_kernel_radius = 5;
+
+    /*
+     * Second pass 🎯
+     * Perform the final thresholding using the dispersion mask.
+    */
+    {
```

Comment:
and this scope REALLY looks like it doesn't do anything?

Reply:
You'd be right https://github.com/DiamondLightSource/fast-feedback-service/commit/e57c8e138c0e1f6edc3d3f35b2df9881bff78895